components/library/RichTextEditor/Preview/RichTextEditor.tsx

```diff
@@ -0,0 +1,135 @@
+import { Box, Icon, InlineGrid, Text } from "@shopify/polaris";
+import { AlertCircleIcon } from "@shopify/polaris-icons";
+import { useEffect, useState } from "react";
+import type { ReactQuillProps } from "react-quill";
+import "react-quill/dist/quill.snow.css";
+import { ClientOnly } from "remix-utils/client-only";
+
+type ReducedQuillProps = Omit<ReactQuillProps, 'readOnly' | 'className' | 'theme' | 'children' | 'style' | 'tabIndex' | 'scrollingContainer'>
+
+const defaultModuleOptions = {
+    toolbar: [
+        ["bold", "italic", "underline", "blockquote"],
+        [
+            { list: "ordered" },
+            { list: "bullet" },
+            { indent: "-1" },
+            { indent: "+1" },
+        ],
+        ["link"],
+        ["clean"],
+    ],
+    clipboard: {
+        matchVisual: false // Needed to avoid this glitch: https://github.com/slab/quill/issues/2905
+    }
+};
+
+export function RichTextEditor({
+    bounds,
+    defaultValue,
+    formats,
+    id,
+    modules,
+    onBlur,
+    onChange,
+    onChangeSelection,
+    onFocus,
+    onKeyDown,
+    onKeyPress,
+    onKeyUp,
+    placeholder,
+    preserveWhitespace,
+    value,
+    label,
+    labelHidden = false,
+    disabled,
+    error,
+    fallbackComponent = <FallbackComponent />
+}: ReducedQuillProps & {
+    label: string;
+    labelHidden?: boolean;
+    disabled?: boolean;
+    error?: string | boolean | undefined;
+    fallbackComponent?: React.ReactNode
+}) {
+
+    const [Quill, setQuill] = useState<any>(null);
+
+    const mergedModuleOptions = {
+        ...defaultModuleOptions,
+        ...modules,
+        toolbar: modules?.toolbar || defaultModuleOptions.toolbar,
+        clipboard: {
+            ...defaultModuleOptions.clipboard,
+            ...(modules?.clipboard || {})
+        }
+    };
+
+
+    useEffect(() => {
```

Comment:
Going off of my other comment about the user handling `ClientOnly` externally, I'm _almost positive_ this won't be needed if you wrap this component in a `ClientOnly` from where it's rendered (i.e. in Example.jsx). It should only ever be loading on the client so there's no need to check if the `window` exists. In which case you could get rid of the `Quill` useState var and condition check on line 89 as well and import Quill directly at the top of the file. 

I would double check after you make these changes by importing to a Remix project and calling this from another component and wrapping in `ClientOnly` where it's called there.

Reply:
I don't think this is correct, because Vite will still import and evaluate quill.js on the server-side, even if it doesn't attempt to render the component. This results in an error in quill.js:
```
ReferenceError: document is not defined
at Object.<anonymous>
(C:\......\node_modules\react-quill\node_modules\quill\dist\quill.js:7661:12
```
As such, you not only need to render client side (<ClientOnly>) but also need to import react-quill on client side.

Reply:
We could potentially move all the logic for loading Quill client side to outside of the component, but as a Remix dev I'd much rather a nice ready-to-go component that doesn't need a bunch of code & imports outside of the RTE component itself in order for it to work.

I may be wrong, but I think ClientOnly can work in non Remix environments as I think the only dependency is React. It might mean this assists with loading in other SSR environments, not just Remix?

Or maybe we need a Remix/SSR version and a purely client-side version?

Reply:
Ah so in addition to wrapping where it's rendered in `ClientOnly`, you have to specify the file as a client-side only file. So naming it `RichTextEditor.client.jsx` will get it to work, just tested in a Remix project. Then I'm using the component on a page route like this:

```
import { BlockStack, Card, Page } from "@shopify/polaris";
import { useState } from "react";
import { RichTextEditor } from "../components/RichTextEditor.client";
import { ClientOnly } from "remix-utils/client-only";

export default function Example() {
  const [textEditor1, setTextEditor1] = useState("");

  return (
    <ClientOnly>
      {() => (
        <Page narrowWidth>
          <Card>
            <BlockStack gap="300">
              <RichTextEditor
                label="Default Editor"
                placeholder="Enter text..."
                onChange={setTextEditor1}
                value={textEditor1}
              />
            </BlockStack>
          </Card>
        </Page>
      )}
    </ClientOnly>
  );
}```

This gets rid of the useEffect/SSR-specific checks within the file to keep the component itself clean, and you can put the Remix changes needed in comments of the file.